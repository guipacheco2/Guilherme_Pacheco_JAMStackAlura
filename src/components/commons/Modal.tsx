import { motion } from 'framer-motion'
import React, { useRef, useState } from 'react'
import styled, { createGlobalStyle, css } from 'styled-components'
import { IconButton } from './IconButton'

interface StyledModalProps {
  isOpen: boolean
}

const StyledModal = styled.div<StyledModalProps>`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  background: rgba(0, 0, 0, 0.1);
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: auto;
  overflow: hidden;
  transition: 0.3s;
  ${({ isOpen }) => {
    if (isOpen) {
      return css`
        opacity: 1;
        pointer-events: all;
      `
    }
    return css`
      opacity: 0;
      pointer-events: none;
    `
  }}
`

interface ModalProps extends StyledModalProps {
  onClose: () => void
  children: React.ReactNode
}

const StyledLockScroll = createGlobalStyle`
  body {
    overflow: hidden;
  }
`

type UseModalOpenHandler = () => void
type UseModalCloseHandler = () => void

export function useModal(
  initialIsOpen = false,
): [boolean, UseModalOpenHandler, UseModalCloseHandler] {
  const [isOpen, setIsOpen] = useState<boolean>(initialIsOpen)

  return [isOpen, () => setIsOpen(true), () => setIsOpen(false)]
}

export function Modal({ isOpen, onClose, children }: ModalProps): JSX.Element {
  const innerElementRef = useRef<HTMLDivElement>(null)

  function handleClickOutsideInnerElement(
    event: React.MouseEvent<HTMLDivElement, MouseEvent>,
  ) {
    if (!innerElementRef.current) throw new Error('Unable to use ref.')

    const isSafeArea = innerElementRef.current.contains(
      event.target as HTMLElement,
    )

    if (!isSafeArea) {
      onClose()
    }
  }

  return (
    <StyledModal isOpen={isOpen}>
      {isOpen && <StyledLockScroll />}
      <motion.div
        onClick={handleClickOutsideInnerElement}
        variants={{ open: { x: 0 }, closed: { x: '100%' } }}
        animate={isOpen ? 'open' : 'closed'}
        transition={{ duration: 0.5 }}
        style={{ display: 'flex', flex: 1 }}
      >
        <div ref={innerElementRef}>
          <div
            style={{
              boxShadow: '-10px 0px 24px rgba(7, 12, 14, 0.1)',
              display: 'flex',
              flexDirection: 'column',
              flex: 1,
              backgroundColor: 'primary',
            }}
          >
            <div style={{ display: 'flex', justifyContent: 'flex-end' }}>
              <IconButton onClick={onClose}>x</IconButton>
            </div>
            <div
              style={{
                display: 'flex',
                flexDirection: 'column',
                justifyContent: 'center',
                flex: 1,
                padding: '16px', // md: 85px
              }}
            >
              {children}
            </div>
          </div>
        </div>
      </motion.div>
    </StyledModal>
  )
}
