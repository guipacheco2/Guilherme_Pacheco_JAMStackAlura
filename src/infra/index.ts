export * from './hooks'
export * from './test'
