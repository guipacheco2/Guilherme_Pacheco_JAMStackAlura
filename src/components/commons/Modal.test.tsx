import React from 'react'
import { render, screen } from '../../infra'
import { Modal } from './Modal'

describe('<Modal />', () => {
  describe('when visibility', () => {
    describe('is closed', () => {
      test('not render content', () => {
        const { baseElement } = render(
          <Modal isOpen={false} onClose={jest.fn()}>
            <div data-testid="content" />
          </Modal>,
        )

        expect(() => screen.getByTestId('content')).toThrow()

        expect(baseElement).toMatchSnapshot()
      })
    })

    describe('is open', () => {
      it('render content', async () => {
        const { baseElement } = render(
          <Modal isOpen={true} onClose={jest.fn()}>
            <div data-testid="content" />
          </Modal>,
        )

        expect(screen.getByTestId('content')).toBeVisible()

        expect(baseElement).toMatchSnapshot()
      })

      it('closes on close button click', async () => {
        const onClose = jest.fn()

        render(
          <Modal isOpen={true} onClose={onClose}>
            <div data-testid="content" />
          </Modal>,
        )

        screen.getByRole('button').click()

        expect(onClose).toBeCalledTimes(1)
      })

      it('closes on overlay element click', async () => {
        const onClose = jest.fn()

        render(
          <Modal isOpen={true} onClose={onClose}>
            <div data-testid="content" />
          </Modal>,
        )

        screen.getByTestId('modal-overlay').click()

        expect(onClose).toBeCalledTimes(1)
      })

      it('ignore close handler on inside element click', async () => {
        const onClose = jest.fn()

        render(
          <Modal isOpen={true} onClose={onClose}>
            <div data-testid="content" />
          </Modal>,
        )

        screen.getByTestId('content').click()

        expect(onClose).toBeCalledTimes(0)
      })
    })
  })
})
