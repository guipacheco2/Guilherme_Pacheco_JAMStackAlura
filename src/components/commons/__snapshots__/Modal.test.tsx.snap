// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> when visibility is closed not render content 1`] = `
.c2 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c5 {
  width: 24px;
  height: 24px;
  color: #34515E;
  background-color: transparent;
  border: none;
  outline: none;
  cursor: pointer;
}

.c3 {
  box-shadow: 0px 1px 25px 5px rgba(0,0,0,0.5);
  border-radius: 10px 10px 0px 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #fff;
  padding: 16px 16px 48px 16px;
  background-color: #FAFAFA;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background: rgba(0,0,0,0.5);
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: auto;
  overflow: hidden;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  opacity: 0;
  pointer-events: none;
}

@media only screen and (min-width:0px) {
  .c2 {
    max-width: initial;
    padding-right: 16px;
    padding-left: 16px;
  }
}

@media only screen and (min-width:600px) {
  .c2 {
    max-width: 600px;
    padding-right: 24px;
    padding-left: 24px;
  }
}

@media only screen and (min-width:1024px) {

}

@media only screen and (min-width:1440px) {

}

@media only screen and (min-width:1920px) {

}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="modal-overlay"
      >
        <div
          class="c2"
        >
          <div
            class="c3"
          >
            <div
              class="c4"
            >
              <button
                class="c5"
              >
                <svg
                  fill="none"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`<Modal /> when visibility is open render content 1`] = `
<body>
  .c2 {
  width: 100%;
  margin-right: auto;
  margin-left: auto;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c5 {
  width: 24px;
  height: 24px;
  color: #34515E;
  background-color: transparent;
  border: none;
  outline: none;
  cursor: pointer;
}

.c3 {
  box-shadow: 0px 1px 25px 5px rgba(0,0,0,0.5);
  border-radius: 10px 10px 0px 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #fff;
  padding: 16px 16px 48px 16px;
  background-color: #FAFAFA;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background: rgba(0,0,0,0.5);
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: auto;
  overflow: hidden;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  opacity: 1;
  pointer-events: all;
}

@media only screen and (min-width:0px) {
  .c2 {
    max-width: initial;
    padding-right: 16px;
    padding-left: 16px;
  }
}

@media only screen and (min-width:600px) {
  .c2 {
    max-width: 600px;
    padding-right: 24px;
    padding-left: 24px;
  }
}

@media only screen and (min-width:1024px) {

}

@media only screen and (min-width:1440px) {

}

@media only screen and (min-width:1920px) {

}

<div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="modal-overlay"
      >
        <div
          class="c2"
        >
          <div
            class="c3"
          >
            <div
              class="c4"
            >
              <button
                class="c5"
              >
                <svg
                  fill="none"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
            <div
              data-testid="content"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
